File: docs/domain/code/ts/src/shared/snapshots.ts
// Shared Kernel — DTO crossing the Vault boundary.
// 真実は Rust 側 (rust/src/snapshots.rs)。
//
// 由来:
//   - glossary.md §3 NoteFileSnapshot, Hydration, ScanFileFailure
//   - aggregates.md §4 Vault Aggregate
//   - domain-events.md `VaultScanned`

import type { FsError } from "./errors.js";
import type { Body, Frontmatter, NoteId, Timestamp } from "./value-objects.js";

/**
 * Vault 境界を渡る DTO。
 * `body` は Markdown 文字列のまま保持する（ファイル境界では Markdown 文字列が一次データ）。
 * Aggregate 境界内では `parseMarkdownToBlocks(body)` で Block[] に変換してから扱う
 * （Hydration の責務、aggregates.md §1.6 / workflows.md Workflow 1）。
 */
export type NoteFileSnapshot = {
  readonly noteId: NoteId;
  readonly body: Body;
  readonly frontmatter: Frontmatter;
  readonly filePath: string;
  readonly fileMtime: Timestamp;
};

/**
 * snapshot → Note Aggregate の変換失敗のみを表す。
 * 由来: glossary.md §3 / domain-events.md `NoteHydrationFailed`
 *
 * read I/O 失敗（permission 等）はここに含めない。
 * scanVault でのファイル単位 read 失敗は `ScanFileFailure.kind = 'read'` を使う。
 */
export type HydrationFailureReason =
  | "yaml-parse"
  | "missing-field"
  | "invalid-value"
  | "block-parse"
  | "unknown";

/**
 * scanVault で個別ファイルが失敗した原因を出所別に表す判別ユニオン。
 * `read` は OS read 失敗（FsError）、`hydrate` は parse / 変換失敗。
 *
 * 由来: workflows.md Workflow 1 Step 2 / domain-events.md `VaultScanned.corruptedFiles`
 */
export type ScanFileFailure =
  | { readonly kind: "read"; readonly fsError: FsError }
  | { readonly kind: "hydrate"; readonly reason: HydrationFailureReason };

export type CorruptedFile = {
  readonly filePath: string;
  readonly failure: ScanFileFailure;
  readonly detail?: string;
};

File: docs/domain/code/rust/src/snapshots.rs
//! Shared Kernel — DTO that crosses the Vault boundary.
//!
//! 由来:
//!   - glossary.md §3 NoteFileSnapshot, Hydration, ScanFileFailure
//!   - aggregates.md §4 Vault Aggregate / NoteSnapshot
//!   - domain-events.md `VaultScanned`

use crate::errors::FsError;
use crate::value_objects::{Body, Frontmatter, NoteId, Timestamp};

// ──────────────────────────────────────────────────────────────────────
// NoteFileSnapshot
// Vault.scan() が返す読み取り表現。Curate 側で Note Aggregate に変換。
// ──────────────────────────────────────────────────────────────────────

/// 永続化された Note の読み取り表現。
/// `Note` Aggregate に変換するのは Curate 側 ACL の責務。
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct NoteFileSnapshot {
    pub note_id: NoteId,
    pub body: Body,
    pub frontmatter: Frontmatter,
    pub file_path: String,
    pub file_mtime: Timestamp,
}

// ──────────────────────────────────────────────────────────────────────
// CorruptedFile — VaultScanned に同梱される失敗ファイル
// domain-events.md `VaultScanned.corruptedFiles`
// ──────────────────────────────────────────────────────────────────────

/// snapshot → Note Aggregate の変換失敗のみを表す。
/// read I/O 失敗はここに含めない（`ScanFileFailure::Read` を使う）。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HydrationFailureReason {
    YamlParse,
    MissingField,
    InvalidValue,
    Unknown,
}

/// scanVault でのファイル単位失敗を出所別に表す判別ユニオン。
/// 由来: workflows.md Workflow 1 Step 2 / domain-events.md `VaultScanned.corruptedFiles`
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ScanFileFailure {
    /// `FileSystem.read_file` が失敗した（permission/lock/not-found 等）。
    Read { fs_error: FsError },
    /// `parser.parse` または snapshot → Note の変換が失敗した。
    Hydrate { reason: HydrationFailureReason },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CorruptedFile {
    pub file_path: String,
    pub failure: ScanFileFailure,
    pub detail: Option<String>,
}
